Extend sssctl to manage cached GPOs

scabrero opened the pull request:
Extends the `sssctl` tool to show, list, remove and purge cached GPOs.

Fixes #4523

thalman reviewed (changes requested):
@scabrero, I reviewd the changes, looks great! Thanks for the effort.

Would you be so kind and update also sss_cache man page? (src/man/sss_cache.8.xml)

Tomáš

thalman commented:
> @scabrero, I reviewd the changes, looks great! Thanks for the effort.
> 
> Would you be so kind and update also sss_cache man page? (src/man/sss_cache.8.xml)
> 
Sorry, mixed up `sssctl` and `sss_cache`. Ignore my comment, please.

T.

thalman reviewed (approved):
LGTM, works as expected, ACK

scabrero commented:
> Hi @scabrero,
> 
> could you please add RN? See https://github.com/SSSD/sssd/blob/master/.git-commit-template#L7 for a hint.

Hi @alexey-tikhonov, I added an empty commit with the :relnote: tag. If you prefer to amend the latest patch just let me know and I will push again.

Thanks for the reviews!

sumit-bose commented on src/tools/sssctl/sssctl_cache.c:
```diff
@@ -713,6 +758,49 @@ errno_t sssctl_netgroup_show(struct sss_cmdline *cmdline,
+
+    ret = parse_cmdline(cmdline, tool_ctx, options, &opts.value, &opts.domain);
+    if (ret != EOK) {
+        ERROR("Failed to parse command line: %s", sss_strerror(ret));
```
Hi,

please add `\n` to the end of `ERROR` messages (e.g. below as well) because they are not added automatically.

bye,
Sumit

sumit-bose commented on src/providers/ad/ad_gpo.c:
```diff
@@ -4923,9 +4923,16 @@ ad_gpo_get_sd_referral_send(TALLOC_CTX *mem_ctx,
     }
 
+    state->smb_host = talloc_strdup(state, lud->lud_host);
+
```
Hi,

if a user has specified a dedicated AD DC for this domain with the `ad_server` option letting `gpo_child` pick a random AD DC from this domain might not be what the user was expecting because e.g. the given AD DC is the only only reachable for the client.

There is a note in the `sssd-ad` man page "Note: Trusted domains will always auto-discover servers even if the primary server is explicitly defined in the ad_server option." which was added 2015 but imo is not true anymore.

If you would move the original code into `ad_gpo_get_sd_referral_conn_done()` the `uri` should be set.

bye,
Sumit

scabrero replied:
This is indeed the right fix, I have tested it and `state->conn->service->uri` is set after `sdap_id_op_connect_{send,recv}`.

sumit-bose commented:
Hi,

thank you for the patches, in general I'm fine but please my in-line comments.

Some general comments. `sssctl gpo-show` and `sssctl gpo.remove` require that the domain name must be given, e.g

    sssctl gpo-show 'Default Domain Policy'@samba.test

or

    sssctl gpo-show -g {31B2F340-016D-11D2-945F-00C04FB984F9}@samba.test

which imo should be mentioned in the `--help` output because it might be unexpected.

Is there is reason you are using the `description` attribute to store the display name and not the `name` attribute? The benefit would be that `name` is already indexed. Additionally, `gpoGUID` isn't indexed as well and since there are now searches by GUID it might be worth to add an index for `gpoGUID` as well?

bye,
Sumit

scabrero commented:
Hi,

I have addressed all suggestions and pushed an update with the following changes:

Hi Sumit,

> Some general comments. `sssctl gpo-show` and `sssctl gpo.remove` require that the domain name must be given
Ok, fixed.
 
> Is there is reason you are using the `description` attribute to store the display name and not the `name` attribute? The benefit would be that `name` is already indexed.

Because I saw that `get_attr_name()` in `sssctl_cache.c` checks `dom->fqnames` to build a FQDN, so I used SYSDB_DESCRIPTION to keep it separated from `use_fully_qualified_names`. I agree with the index reasoning so I have changed it to use SYSDB_NAME but kept a separate handler for it.

> Additionally, `gpoGUID` isn't indexed as well and since there are now searches by GUID it might be worth to add an index for `gpoGUID` as well?

Yes, and I made it case-insensitive for searches.

Summary of changes:
- Switch to Sumit's suggested fix for GPO referral processing when subdomain connection is not yet initialized.
- Store the AD displayName attribute in SYSDB_NAME instead of SYSDB_DESCRIPTION.
- Bump sysdb version to 0.24: Index gpoGUID and make it case-insensitive on searches.
- Add new parameter to show extended help for `sssctl` subcommands
- Add extended help for `gpo-show` and `gpo-remove` subcommands.

sumit-bose commented on src/tools/sssctl/sssctl_cache.c:
```diff
@@ -713,6 +763,505 @@ errno_t sssctl_netgroup_show(struct sss_cmdline *cmdline,
+        goto done;
+    }
+
+    ret = get_attr_string(tmp_ctx, entry, dom, SYSDB_DESCRIPTION, &gpo_name);
```
Hi,

I guess this should be `SYSDB_NAME` now.

bye,
Sumit

sumit-bose commented on src/tools/sssctl/sssctl_cache.c:
```diff
@@ -713,6 +763,505 @@ errno_t sssctl_netgroup_show(struct sss_cmdline *cmdline,
+        goto done;
+    }
+
+    attr = SYSDB_DESCRIPTION;
```
Same, here `SYSDB_NAME`.

sumit-bose commented on src/tools/sssctl/sssctl_cache.c:
```diff
@@ -718,6 +763,55 @@ errno_t sssctl_netgroup_show(struct sss_cmdline *cmdline,
+    const char *attr;
+    errno_t ret;
+    const char *extended_help =
+        "This command requires the domain name to be given when the "
```
Hi,

I think this is currently not correct. In my tests I have always provide the domain name even if there is only a single domain. So either the help message should be changes to something like `... requires the domain name to be given because the same policy name (or GUID) might exists ...` or the code should be changed. I would prefer to just change the help message an always require the domain name to be given to make the usage more consistent.

bye,
Sumit

scabrero replied:
Makes sense, domain is now mandatory.

sumit-bose reviewed (approved):
Hi,

thank you for your patience, I have no further comments, ACK.

bye,
Sumit

alexey-tikhonov commented:
Hi @scabrero,

could you please rebase on the top of latest 'master' branch?

pbrezina commented:
* `master`
  * 54179a094f732eaadf925bd869d3097889f4a213 - SSSCTL: Add the new cached GPOs management commands to release notes
  * c5b16eec476614133f9a6643015d5f0094440e0b - SSSCTL: Add gpo-purge command
  * afee68b113b4cab7a145de83c798f1e0cd34841c - SSSCTL: Add sssctl gpo-remove command
  * be735999d221481302d8dd07192b3a6e2893d105 - SYSDB: Add a function to delete GPO entry by GPO GUID
  * 6dc9166c2c04bf5d71ddc7a08de49a3b022a1212 - SSSCTL: Add sssctl gpo-list command
  * 18a17bcd506b70412b6f29ce7dc06828c4089f55 - SSSCTL: Add gpo-show command
  * 095e31eb23dd1fb845fb928a345d4c204284b3f1 - SSSCTL: Prepare for extended help in subcommands
  * cf59da1aa7f357b66cd4baaf6566c5c2d42ae3a7 - SYSDB: Add new index for gpoGUID and make searches on it case insensitive
  * 66fd8a04891be03f836256216e1fe7c221d7a0ee - SYSDB: Always canonicalize GPO guid
  * 35801347e270fa515e5f54068b4b4de773001325 - SYSDB: Store the GPO's filesystem path in sysdb entry
  * 568ca5dee4efa15cace0e6ff3a7c04021c48f56e - SYSDB: Store GPO's displayName in sysdb
  * e1692772bf6bed9062a77518e9c0dc66dfc51942 - GPO: Fetch the GPO's displayName attribute
  * 9926067110e8fd26943dfe1efe1ed34b54735ae3 - SYSDB: Add sysdb_gpos_base_dn()
  * 98efb5ec92ed28dc9e74821526dfc773c3ca8044 - GPO: Remove unused local variable
  * 738bb53307ee90fccb97be6d62c850abdc259c03 - GPO: Defer SMB server choice until id connection established when processing referrals
